src/hotspot/share/gc/g1/g1FullGCPrepareTask.cpp

```diff
@@ -49,8 +49,15 @@ bool G1FullGCPrepareTask::G1CalculatePointersClosure::do_heap_region(HeapRegion*
       if (!_bitmap->is_marked(obj)) {
         free_humongous_region(hr);
       }
+    } else if (hr->is_open_archive()) {
+      bool is_empty = _bitmap->get_next_marked_addr(hr->bottom(), hr->top()) >= hr->top();
+      if (is_empty) {
+        free_pinned_region(hr);
+      }
     } else {
-      assert(hr->is_archive(), "Only archive regions can also be pinned.");
+      // Never free closed archive regions. This is also be the only other allowed
+      // type at this point.
+      assert(hr->is_closed_archive(), "Only closed archive regions can also be pinned.");
```

Comment:
I found the assert message here very confusing.  It's really that all other pinned region cases have been covered, and closed_archive is the last remaining one.